### DISCUSSION TITLE
How do I reuse a scheme as a nested field of another type

### DISCUSSION THREAD

**Discussion by avnerbarr** [colinhacks/zod#2812] [Q&A] [answered]
trying to achieve this:

```ts
type A {
   field1: string;
   field2: number;
}

type B {
  arrayOfA: A[];
}
```

I'm not clear on the syntax to achieve that:

```ts
// equivalent to type A
const AZod = z.object({
field1: z.string(),
field2: z.number()
});


// equivalent to type B
const BZod = z.object(z.array(z.infer<typeof AZod>) // is this correct?
```

also, wondering if it is possible to skip this entirely and just infer directly from the types I already have?

Want to do this:

```ts
type A {
   field1: string;
   field2: number;
}

// type A schema zod validator directly from the type
const AZod = z.Zod<A> // create the A zod type without typing everything?

type B { ... } 
const BZod = z.Zod<B> // ?
```

**Comment by dgritsko** [comment]
> wondering if it is possible to skip this entirely and just infer directly from the types I already have?

No, it's not, but you can go the other way around (start with your Zod schemas, and then infer the types using `z.infer`). I think this is what you're looking for:

```ts
const AZod = z.object({
  field1: z.string(),
  field2: z.number()
});

type A = z.infer<typeof AZod>;

const BZod = z.object({ arrayOfA: z.array(z.string()) });

type B = z.infer<typeof BZod>;
```

**Comment by avnerbarr** [comment]
thanks!

Need clarification:

In the snippet you wrote:

```ts
const BZod = z.object({ arrayOfA: z.array(z.string()) }); 
```

but I want `arrayOfA: A[] // how is the "A" type written here?` - what is the correct syntax for this?  perhaps `A[]` or maybe `arrayOfA: z.infer<typeof AZod>`? 


Would be great to provide this type of example in the documentation or point me to something like that if it already exists (I didn't find how to nest existing types)

**Comment by dgritsko** [comment, marked as answer]
You'd just use your `AZod` schema with `z.array`, like this:

```ts
const BZod = z.object({ arrayOfA: z.array(AZod) });
```
